syntax = "proto3";
option java_package = "grpc.server.services";


/*  file transfer */
message UploadRequest {
  oneof data {
    string filename = 2;
    bytes chunk_data = 3;
  }
}

message UploadReply {
  string filename = 1;
}

/* PRISM engine */

message PrismDevNullLog {}

message PrismFileLog {
  string type = 1;
}

message PrismLog {
  oneof log_type {
    PrismDevNullLog dev_null_log = 1;
    PrismFileLog file_log = 2;
  }
}

message InitialiseRequest {
  string prism_object_id = 1;
}

message InitialiseResponse {
  string status = 2;
}

message LoadPRISMModelRequest {
  string prism_object_id = 1;
  string module_object_id = 2;
}

message LoadPRISMModelResponse {
  string prism_object_id = 1;
  string module_object_id = 2;
  string status = 3;
}


message ParsePropertiesFileRequest {
  string prism_object_id = 1;
  string module_object_id = 2;
  string properties_file_object_id = 3;
  string properties_file_name = 4;
}

message ParsePropertiesFileResponse {
  string prism_object_id = 1;
  string status = 2;
  string properties = 3;
}




message ModelCheckRequest {
  string prism_object_id = 1;
  string properties_file_object_id = 2;
  string property_object_id = 3;
  string result_object_id = 4;
}

message ModelCheckResponse {
  string result = 1;
  string status = 2;
}










message ParseAndLoadModelRequest {
  string prism_object_id = 1;
  string model_file_name = 2;
}

message ParseAndLoadModelReply {
  string prism_object_id = 1;
  string status = 2;
}

message ClosePrismRequest {
  string prism_object_id = 1;
}

message ClosePrismResponse {
  string prism_object_id = 1;
  string status = 2;
}

message ParseModelFileRequest {
  string prism_object_id = 1;
  string module_object_id = 2;
  string model_file_name = 3;
}

message ParseModelFileResponse {
  string prism_object_id = 1;
  string module_object_id = 2;
  string status = 3;
}

message GetPropertyObjectRequest {
  string properties_file_object_id = 1;
  string property_object_id = 2;
  int32 property_index = 3;
}

message GetPropertyObjectResponse {
  string properties_file_object_id = 1;
  string status = 2;
  string property = 3;
}

message PropertiesFileForwardMethodCallRequest {
  string properties_file_object_id = 1;
  string method = 2;
  string args = 3;
  string kwargs = 4;
}

message PropertiesFileForwardMethodCallResponse {
  string properties_file_object_id = 1;
  string status = 2;
  string result = 3;
}
message GetUndefinedConstantsUsedInPropertyRequest {
  string properties_file_object_id = 1;
  string property_object_id = 2;
}

message GetUndefinedConstantsUsedInPropertyResponse {
  string status = 1;
  repeated string constants = 2;
}

message AddValueRequest {
  string values_object_id = 1;
  string const_name = 2;
  int32 value = 3;
}

message AddValueResponse {
  string result = 1;
  string status = 2;
}

message SetSomeUndefinedConstantsRequest {
  string properties_file_object_id = 1;
  string values_object_id = 2;
}

message SetSomeUndefinedConstantsResponse {
  string status = 1;
}

message DeleteObjectRequest {
  string object_id = 1;
}

message DeleteObjectResponse {
  string status = 1;
}


message InitUndefinedConstantsResponse {
  string status = 1;
}

message DefineUsingConstSwitchRequest{
  string undefined_constants_object_id = 1;
  string constant = 2;
}

message DefineUsingConstSwitchResponse{
  string status = 1;
}

message GetNumberPropertyIterationsRequest {
  string undefined_constants_object_id = 1;
}

message GetNumberPropertyIterationsResponse {
  string status = 1;
  int32 number_iterations = 2;
}


message InitValuesResponse {
  string status = 1;
}

message GetPFConstantValuesRequest {
  string undefined_constants_object_id = 1;
  string values_object_id = 2;
}

message GetPFConstantValuesResponse {
  string status = 1;
  string values = 2;
}

///////////////////////////////////////// Initializer Messages /////////////////////////////////////////
// Init response (for all initializers)
message InitResponse {
  string status = 1;
}
// Init arguments
message InitPrismLogRequest {
  string prism_log_object_id = 1;
  string type = 2;
}

message InitPrismRequest {
  string prism_object_id = 1;
  string main_log_object_id = 2;
}

message InitModulesFileRequest {
  string modules_file_object_id = 1;
}



message InitPrismDevNullLogRequest {
  string main_log_object_id = 1;
}

message InitPrismFileLogRequest {
  string main_log_object_id = 1;
  string output_type = 2;
}

message InitPropertiesFileRequest {
  string properties_file_object_id = 1;
}

message InitPropertyObjectRequest {
  string property_object_object_id = 1;
}

message InitResultRequest {
  string result_object_id = 1;
}


message InitUndefinedConstantsRequest {
  string module_object_id = 1;
  string properties_file_object_id = 2;
  string property_object_id = 3;
  string undefined_constants_object_id = 4;
}

message InitValuesRequest {
  string values_object_id = 1;
}



// Service Definitions
service PrismProtoService {
  ///////////////////////////////////////// Initializer /////////////////////////////////////////
  // Prism Log
  rpc InitPrismLog(InitPrismLogRequest) returns (InitResponse) {}
  // ModulesFile
  rpc InitModulesFile(InitModulesFileRequest) returns (InitResponse) {}
  // Prism
  rpc InitPrism(InitPrismRequest) returns (InitResponse) {}
  // PropertiesFile
  rpc InitPropertiesFile(InitPropertiesFileRequest) returns (InitResponse) {}
  // Property Object (called Property Object because "property" is a reserved word in Python)
  rpc InitPropertyObject(InitPropertyObjectRequest) returns (InitResponse) {}
  // Result
  rpc InitResult(InitResultRequest) returns (InitResponse) {}
  // init undefined constants
  rpc InitUndefinedConstants(InitUndefinedConstantsRequest) returns (InitResponse) {}
  // Values
  rpc InitValues(InitValuesRequest) returns (InitResponse) {}

  ///////////////////////////////////////// Generic Functions /////////////////////////////////////////
  // delete a PRISM object
  rpc DeleteObject(DeleteObjectRequest) returns (DeleteObjectResponse) {}
  // Generic method to upload files
  rpc UploadFile (stream UploadRequest) returns (UploadReply) {}

  ///////////////////////////////////////// Object specific functions /////////////////////////////////////////
  // Initialise the PRISM engine
  rpc Initialise(InitialiseRequest) returns (InitialiseResponse) {}
  // Parse model file
  rpc ParseModelFile(ParseModelFileRequest) returns (ParseModelFileResponse) {}
  // Load a PRISM model
  rpc LoadPRISMModel(LoadPRISMModelRequest) returns (LoadPRISMModelResponse) {}
  // Parse a properties file
  rpc ParsePropertiesFile(ParsePropertiesFileRequest) returns (ParsePropertiesFileResponse) {}
  // Get propety object from properties file
  rpc GetPropertyObject(GetPropertyObjectRequest) returns (GetPropertyObjectResponse) {}
  // Model check
  rpc ModelCheck(ModelCheckRequest) returns (ModelCheckResponse) {}
  // Get undefined constants used in property
  rpc GetUndefinedConstantsUsedInProperty(GetUndefinedConstantsUsedInPropertyRequest) returns (GetUndefinedConstantsUsedInPropertyResponse) {}
  // add values
  rpc AddValue(AddValueRequest) returns (AddValueResponse) {}
  // set some undefined constants
  rpc SetSomeUndefinedConstants(SetSomeUndefinedConstantsRequest) returns (SetSomeUndefinedConstantsResponse) {}
  // define using const switch for undefined constants
  rpc DefineUsingConstSwitch(DefineUsingConstSwitchRequest) returns (DefineUsingConstSwitchResponse) {}
  // get number of property iterations for undefined constants
  rpc GetNumberPropertyIterations(GetNumberPropertyIterationsRequest) returns (GetNumberPropertyIterationsResponse) {}
  // get properties file constant values
  rpc GetPFConstantValues(GetPFConstantValuesRequest) returns (GetPFConstantValuesResponse) {}

}
